include/core/Concurrency.h

```diff
@@ -146,62 +130,66 @@ using future = concurrency_detail::CTypedFutureAnyWrapper<R>;
 //! them. Prefer using high level primitives, such as parallel_for_each, which are
 //! safer.
 template<typename FUNCTION, typename... ARGS>
-future<std::result_of_t<std::decay_t<FUNCTION>(std::decay_t<ARGS>...)>>
+std::future<std::result_of_t<std::decay_t<FUNCTION>(std::decay_t<ARGS>...)>>
 async(CExecutor& executor, FUNCTION&& f, ARGS&&... args) {
     using R = std::result_of_t<std::decay_t<FUNCTION>(std::decay_t<ARGS>...)>;
 
     // Note g stores copies of the arguments in the pack, which are moved into place
     // if possible, so this is safe to invoke later in the context of a packaged task.
     auto g = std::bind<R>(std::forward<FUNCTION>(f), std::forward<ARGS>(args)...);
 
-    std::packaged_task<boost::any()> task([g_ = std::move(g)]() mutable {
-        return concurrency_detail::resultToAny(g_, std::is_same<R, void>{});
-    });
-    auto result = task.get_future();
+    auto promise = std::make_shared<std::promise<R>>();
+    auto result = promise->get_future();
+
+    std::function<void()> task{[g_ = std::move(g), promise_ = std::move(promise)]() mutable {
+        concurrency_detail::invokeAndWriteResultToPromise(g_, promise_,
+                                                          std::is_same<R, void>{});
+}
+};
 
-    // Schedule the task to compute the result.
-    executor.schedule(std::move(task));
+// Schedule the task to compute the result.
+executor.schedule(std::move(task));
 
-    return std::move(result);
+return std::move(result);
```

Comment:
The indenting looks a bit off here and in the 4 lines above - is what clang-format decided?

Reply:
Yes. I've seen this before when using brace initialiser with a lambda. I'll rejig and see if it improves the formatting.